### PR TITLE
Adding a TypeScript session

### PR DESCRIPTION
**Pull request by s-nicholson**
This adds a TypeScript session.

Starting with hello world, adding in some stuff about type checking and then classes, interfaces, enums and a few other bits.
Any suggestions or contributions would be magic 😄

### CODE REVIEW

**Review by hasib-ahmed1 — COMMENTED** [review]
LGTM 👍 Not sure if I can approve

File: sessions/TypeScript-100/101/101.001/app.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function hello(name: string) {
+    console.log(`Hello, ${name}!`);
+}
+
+hello("World");
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.001/app.ts:5]: Would be nicer to show the possibility of setting the output type:

```suggestion
function hello(name: string): string {
    return `Hello, ${name}!`;
}

console.log(hello("World"));
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+5 halve("some string");
+        ~~~~~~~~~~~~~
+```
+
+## A bit more complicated
+The compiler makes working with more complex types easier by checking what properties they have.
+
+For example, accessing a property that the object doesn't have, gives a type error:
+```
+let obj = {
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:35]:
```suggestion
const obj = {
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+## A tiny bit more complicated again
+You can use this kind of complex type checking in your function definitions as well. 
+
+This lets you say something like _"this function accepts an object which has a `size` property which is a `number`"_:
+```
+function printSize(obj: {size: number}) {
+    console.log(obj.size);
+}
+
+let x = { size: 10 };
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:52]:
```suggestion
const x = { size: 10 };
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+printSize(x); // 10
+
+let y = { size: "10" };
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:55]:
```suggestion
const y = { size: "10" };
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+printSize(y); // compile error: Type 'string' is not assignable to type 'number'.
+
+let z = { age: 42 };
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:58]:
```suggestion
const z = { age: 42 };
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+printSize(z); // compile error: Property 'size' is missing in type '{ age: number; }' but required in type '{ size: number; }'.
+```
+
+You can also use this to check the function properties an object has:
+```
+function callSayHello(
+    obj: {sayHello(name: string): string}
+) {
+    let name: string = "world";
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:67]:
```suggestion
    const name: string = "world";
```

File: sessions/TypeScript-100/101/101.002/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    let greeting: string = obj.sayHello(name); // call function on the object
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:68]:
```suggestion
    const greeting: string = obj.sayHello(name); // call function on the object
```

File: sessions/TypeScript-100/101/101.002/app.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+/*
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/app.ts:1]: Please change all `let` s to `const`s

File: sessions/TypeScript-100/101/101.002/README.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    console.log(greeting);
+}
+
+callSayHello({
+    // object knows how to say hello
+    sayHello(name: string): string {
+        return `hello ${name}`;
+    }
+});
```
Comment (ehsankhfr) [inline sessions/TypeScript-100/101/101.002/README.md:77]: This example might be a bit confusing for a beginner.

**Review by ehsankhfr — COMMENTED** [review]
Great work! just had a few comments around coding style and possible improvement in examples.

File: sessions/TypeScript-100/101/101.002/README.md (hunk 9 above)
Comment (s-nicholson) [reply sessions/TypeScript-100/101/101.002/README.md:77]: Agreed, would probably be a clearer example with the `type` broken out.
What do you think?
```suggestion
/*
    Defines a type of object which has a `sayHello` function which
    accepts a string and returns a string.
*/
type CanSayHello = {
    sayHello(name: string): string
}

const obj: CanSayHello = {
    sayHello(name: string): string {
        return `hello ${name}`;
    }
};

callSayHello(obj);
```

**Review by ehsankhfr — APPROVED** [review]

**Pull request merged**